### PR TITLE
FileIcon addition as Element; asSvg prop added to Icon factories

### PR DESCRIPTION
**Pull request by petradish**
Hoist P/R Checklist
-------------------

**Pull request authors:** Review and check off the below. Items that do not apply can also be
checked off to indicate they have been considered. If unclear if a step is relevant, please leave
unchecked and note in comments.

- [x] Caught up with `develop` branch as of last change.
- [x] Added CHANGELOG entry, or determined not required.
- [x] Reviewed for breaking changes, added `breaking-change` label + CHANGELOG if so.
- [x] Updated doc comments / prop-types, or determined not required.
- [x] Reviewed and tested on Mobile, or determined not required.
- [x] Created Toolbox branch / PR, or determined not required. 

**If your change is still a WIP**, please use the "Create draft pull request" option in the split
button below to indicate it is not ready yet for a final review.

> **Pull request reviewers:** when merging this P/R, please consider using a *squash commit* to
> collapse multiple intermediate commits into a single commit representing the overall feature
> change. This helps keep the commit log clean and easy to scan across releases. PRs containing a
> single commit should be *rebased* when possible.

This PR addresses https://github.com/xh/hoist-react/issues/1513
- New fileIcon Element and Component created to set default configuration (tooltip name and color) for file extensions.
 - Default configs are overridable by passing in add'tl props
- asSvg prop added to all Icon elements -- no longer require the function convertIconToSvg to render icons within AgGrid renderers.

### CODE REVIEW

File: cmp/grid/Grid.js [hunk 1]
```diff
@@ -349,7 +343,7 @@ class LocalModel {
 
             let icon = displaySpec.icon;
             if (isValidElement(icon)) {
```
Comment (lbwexler) [inline cmp/grid/Grid.js]: This seems wrong to me.  Think you need to test this.

Comment (lbwexler) [reply cmp/grid/Grid.js]: (Probably want to leave this one as is.)

File: desktop/cmp/grid/columns/Actions.js [hunk 2]
```diff
@@ -70,7 +69,7 @@ export const actionCol = {
             if (!isEmpty(tooltip)) buttonEl.setAttribute('title', tooltip);
             if (!isEmpty(intent)) buttonEl.classList.add(`bp3-intent-${intent}`);
 
-            buttonEl.innerHTML = convertIconToSvg(icon);
```
Comment (lbwexler) [inline desktop/cmp/grid/columns/Actions.js]: again problematic here -- my guess is that you need to revert this code to get the html you need here.

File: icon/Icon.js [hunk 3]
```diff
@@ -915,43 +925,66 @@ export const deserializeIcon = function(iconDef) {
  * @param {String} extension
  * @return {Element}
  */
-export const fileIcon = function(extension) {
+
+export const [FileIcon, fileIcon] = hoistCmp.withFactory({
```
Comment (lbwexler) [inline icon/Icon.js]: Think we can skip the component, since we have Icon.fileIcon()

**Comment by lbwexler** [comment]
Note: this is no longer a breaking change if we leave convertIconToSvg() in place, also, I am suggesting we don't need the component FileIcon/fileIcon, if we have the elementFactory Icon.fileIcon already in place.

Flagging @amcclain for your thoughts -- 

We have the strange situation where the hoist API passes these "Icons" around as react elements. We always create them with our icon element factories, e.g. . Icon.fooBar(), but there is nothing in the API that really requires that.  For the most part, they are just rendered in react, and could in fact be arbitrary react elements.  But in certain circumstances -- e.g. creating menus, or passing to Golden Layout, we need to get the svg representation, and make a hacky dive into the elements property to pull out the underlying 

Seems like there are two alternatives to how to better document/improve this.

1) Have an explicit Icon Object as the type of our various API properties, rather than Element (either our own, or FontAwesome's).  This would be something that could always be guaranteed to be turned in to either an react element, or SVG.  This might be more backward compatible than we think, but might also be confusing when users interact directly with say, blueprint, and need to call an additional .element() method on objects.  

2)Double down on our current assumption. Just improve our documentation regarding "Icon".  Maybe add a JSDoc for `IconElement` to indicate the special type of element we are passing around.    Add a throwIf() 

3) Stay flexible and be robust. Allow users to pass any React element as an icon.  If we ever need to get an SVG out of it, and can't, log an error and return some default icon....

File: cmp/grid/Grid.js (hunk 1 above)
Comment (petradish) [reply cmp/grid/Grid.js]: It works in toolbox!

File: desktop/cmp/grid/columns/Actions.js (hunk 2 above)
Comment (petradish) [reply desktop/cmp/grid/columns/Actions.js]: This also works if RecordAction takes an icon that has config {asSvg: true}. Not sure how to document that change more clearly (and also why this is a breaking change).
See RecordAction.js line 46: ` // @param {Element} [c.icon] - icon to be displayed. Always specify with config {asSvg: true}.`

**Comment by lbwexler** [comment]
John nice idea about adding the method to the element, but not gonna work.  React elements are immutable and frozen solid like the lake you were on yesterday

File: styles/XH.scss [hunk 4]
```diff
@@ -98,6 +98,16 @@ body.xh-app {
 
   .xh-yellow-light { color: var(--xh-yellow-light); }
 
+  .xh-fileIcon-excel { color: #0e7b41; }
```
[the remark this answers is not in the document]
Comment (petradish) [reply styles/XH.scss]: Thanks for the catch on this and the word .doc extension class name! I am not sure when a color warrants it's own variable. I'll leave it up to the big cheeses for that decision.

**Comment by TomTirapani** [comment]
Took a look at this - think it's a great change. I think we should make sure we highlight the new `asSvg` option in the CHANGELOG, and encourage devs to use that instead of `convertIconToSvg()`

**Comment by lbwexler** [comment]
So John and Tom and I discussed this, and were all leaning toward option 3) above.  It seems like its a pretty important/useful feature that generally our 'icon' props accept elements are are not in anyway tied to Font-Awesome or our own classes.  The element expectation makes the most sense in general.

The new asSvg option on our Elem factory is a nice to  

Will catch this up, and enhance the changelog a bit and look to merge.  Tom will also review.

File: styles/XH.scss (hunk 4 above)
[the remark this answers is not in the document]
Comment (lbwexler) [reply styles/XH.scss]: lets just leave them hard-coded but please fix the camel case

File: icon/Icon.js (hunk 3 above)
Comment (lbwexler) [reply icon/Icon.js]: let's remove this component -- users can just use the elem factory in their render method.

File: desktop/cmp/grid/columns/Actions.js (hunk 2 above)
Comment (lbwexler) [reply desktop/cmp/grid/columns/Actions.js]: Ah -- I see, but then we are lying with jsdoc {Element} declaration on line 46 -- its not an element, its a string of html.   This is a tricky bit, lets dicsuss tommorrow.

**Pull request closed without merging**